src/nupic/algorithms/Connections.cpp

```diff
@@ -471,6 +471,7 @@ void Connections::raisePermanencesToThreshold(
     return;
 
   vector<Synapse> &synapses = segData.synapses;
+  if( synapses.empty()) return; //otherwise crashes! can empty segments/synapses naturally occur? 
```

Comment:
this is the quick fix

Reply:
I don't think the problem is that `synapses` is empty, but rather that `minPermSynPtr` is not between `synapses.begin()` & `synapses.end()`.   In debug mode, the iterator code will check these bounds and throw an error.  In release mode, it might dereference the iterator and segfault.

The situation happens when synapses are pruned so the segment has too few synapses to ever activate.  In this case the method should do the next best thing and connect as many synapses as it can.
`segmentThreshold = min( segmentThreshold, synapses.size() )`

The user probably should  have removed the segment, but the connections class should not make that decision.  Maybe the caller has multiple threshold, such as with the TM?  In the TM: a segment that does not meet the activation threshold can still meet the matching/learning threshold.

Reply:
>  synapses is empty, but rather that minPermSynPtr is not between synapses.begin() & synapses.end() 

This was my problem what to do. Pointer does not fit because threshold > size(synapses). 

Now, should we? 
- complain synapses are empty (too few)? 
- complain thresh is too large? 
- silently abort (return)
- crop thresh to somehow fit? 
  - still what to do with zero-sized synapses? Threshold cannot raise any synapses then. 

> he situation happens when synapses are pruned so the segment has too few synapses to ever activate. In this case the method should do the next best thing and connect as many synapses as it can.
segmentThreshold = min( segmentThreshold, synapses.size() )

In the cases I observed this was (only?) when synapses==0, so the min() would not help the situation. 

> The user probably should have removed the segment, but the connections class should not make that decision. Maybe the caller has multiple threshold, such as with the TM? In the TM: a segment that does not meet the activation threshold can still meet the matching/learning threshold. 

This was my idea, should we prune? Do we distinguis Segments that are emptied and should be removed, and those that just too many synapses were removed? 

As https://discourse.numenta.org/t/temporalmemory-running-very-slow-after-long-training/5052 
I suspect segments, synapses are not removed/repurposed. I'll add a testcase and issue for this later.

Reply:
> In the cases I observed this was (only?) when synapses==0

An iterator which is outside of the vector's data range is invalid and we should avoid it, even if it does not crash or cause problems.

> should we prune?

No. The SP calls this method, but the SP does not do any pruning.  The TM already has code to do pruning, but it doesn't ever call this method.

Reply:
ok, so thresh = min(thresh, synapses.size)
if thresh == 0 -> return early.